### PR TITLE
ignore invalid frames on closed streams

### PR DESCRIPTION
**Pull request by geoffkizer**
Contributes to #38913 
Contributes to #35652 

Ignore any frames received on a closed stream -- more specifically, on a stream that is no longer in the stream dictionary. 

We don't distinguish between streams closed because they were reset by the client (due to cancellation etc) or streams closed because they completed. The former is valid, while the latter is technically invalid per RFC. However, to avoid having to track why the stream was closed, we just ignore these frames in both cases. This matches WinHttp behavior.

We are still not handling the case where a stream is cancelled during the processing of a HEADER or DATA frame.

@dotnet/ncl

### CODE REVIEW

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs [hunk 1]
```diff
@@ -341,9 +339,9 @@ private async Task ProcessHeadersFrame(FrameHeader frameHeader)
...
             _hpackDecoder.CompleteDecode();
 
-            http2Stream.OnResponseHeadersComplete(endStream);
+            http2Stream?.OnResponseHeadersComplete(endStream);
 
-            if (endStream)
+            if (endStream && http2Stream != null)
             {
                 RemoveStream(http2Stream);
             }
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs]: Nit: just to save a null check (and maybe make the logic a little more clear), maybe instead make this:
```C#
if (http2Stream != null)
{
    httpStream.OnResponseHeadersComplete(endStream);
    if (endStream)
    {
        RemoveStream(http2Stream);
    }
}
```
?

**Comment by geoffkizer** [comment]
I pushed another change to also ignore frame processing callbacks on aborted streams. This seems to make all the stress issues here go away. Please take a look.

**Comment by geoffkizer** [comment]
I pushed one additional change: The connection-level window accounting on cancelled streams was incorrect. It would not account for frames received after cancellation, meaning that over time the window would effectively lose credit and shrink, eventually to 0. 

The fix is to just directly account for the connection-level window in ProcessDataFrame. This has the additional benefit of being simpler and more obviously correct.

File: src/Common/tests/System/Net/Http/Http2LoopbackConnection.cs [hunk 2]
```diff
@@ -565,7 +565,11 @@ public async Task PingPong()
         {
             PingFrame ping = new PingFrame(new byte[8] { 1, 2, 3, 4, 50, 60, 70, 80 }, FrameFlags.None, 0);
             await WriteFrameAsync(ping).ConfigureAwait(false);
-            await ReadFrameAsync(Timeout).ConfigureAwait(false);
+            Frame pingAck = await ReadFrameAsync(Timeout).ConfigureAwait(false);
+            if (pingAck.Type != FrameType.Ping || !pingAck.AckFlag)
```
Comment (krwq) [inline src/Common/tests/System/Net/Http/Http2LoopbackConnection.cs]: nit: something we might want to add is also checking if bytes match

File: src/Common/tests/System/Net/Http/Http2LoopbackConnection.cs [hunk 3]
```diff
@@ -565,7 +565,11 @@ public async Task PingPong()
...
+            {
+                throw new Exception("Expected PING ACK");
```
Comment (davidsh) [inline src/Common/tests/System/Net/Http/Http2LoopbackConnection.cs]: Why is it throwing `Exception`?  Usually these are `Http2ProtocolException` with a suitable error number etc.

Comment (geoffkizer) [reply src/Common/tests/System/Net/Http/Http2LoopbackConnection.cs]: It's just test code.

**Pull request merged**